keps/sig-node/4009-add-cdi-devices-to-device-plugin-api/README.md

```diff
@@ -0,0 +1,853 @@
+
+To a vendor changes in the API should always be backwards compatible.
+
+### Version Skew Strategy
+
+Kubelet will always be backwards compatible, so going forward existing plugins are not expected to break.
+<!--
+If applicable, how will the component handle version skew with other
+components? What are the guarantees? Make sure this is in the test plan.
+
+Consider the following in developing a version skew strategy for this
+enhancement:
+- Does this enhancement involve coordinating behavior in the control plane and
+  in the kubelet? How does an n-2 kubelet without this feature available behave
+  when this feature is used?
+- Will any other components on the node change? For example, changes to CSI,
+  CRI or CNI may require updating that component before the kubelet.
+-->
+
+## Production Readiness Review Questionnaire
+
+<!--
+
+Production readiness reviews are intended to ensure that features merging into
+Kubernetes are observable, scalable and supportable; can be safely operated in
+production environments, and can be disabled or rolled back in the event they
+cause increased failures in production. See more in the PRR KEP at
+https://git.k8s.io/enhancements/keps/sig-architecture/1194-prod-readiness.
+
+The production readiness review questionnaire must be completed and approved
+for the KEP to move to `implementable` status and be included in the release.
+
+In some cases, the questions below should also have answers in `kep.yaml`. This
+is to enable automation to verify the presence of the review, and to reduce review
+burden and latency.
+
+The KEP must have a approver from the
+[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
+team. Please reach out on the
+[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
+you need any help or guidance.
+-->
+
+### Feature Enablement and Rollback
+
+<!--
+This section must be completed when targeting alpha to a release.
+-->
+
+###### How can this feature be enabled / disabled in a live cluster?
+
+<!--
+Pick one of these and delete the rest.
+
+Documentation is available on [feature gate lifecycle] and expectations, as
+well as the [existing list] of feature gates.
+
+[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
+[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
+-->
+
+This change is not behind a feature gate. It is adding an optional
+field to a stable API. This mirrors what was done in [KEP-1884](https://github.com/kubernetes/enhancements/pull/1884) when
+adding a `cpu_ids` field to the existing PodResources API.
+
+<!-- Furthermore, the only code that would be featured-gated in the Device Manager would be the code to propagate this field from the `AllocateResponse` to the
+corresponding field in the CRI. -->
```

Comment:
This does not sound right to me.  This new field should start as alpha with its own feature gate.  It does not matter that the field is optional.

Reply:
Agreed @enj. KEP-1884 was prior to requiring PRR.

We should feature gate this and use standard feature rollout processes. The API you are adding to may not be alpha, but the field still needs a feature gate.

Reply:
Note that the field itself can't be feature gated (it's a field in a prototbuf), and the code to add the feature gate will actually require more code than the code required to pass the value of the field on to the CRI (where the exact same field with the exact same format already exists).

It seems overly complicated in my opinion to make such a small change go through the standard alpha/beta/stable graduation criteria.

Reply:
> the code to add the feature gate will actually require more code than the code required to pass the value of the field on to the CRI

Feature gating always adds branching and complexity.  This is normal and expected.

> It seems overly complicated in my opinion to make such a small change go through the standard alpha/beta/stable graduation criteria.

See #2784 as an example of going through the entire KEP process for adding a simple (new/optional) field to a GA API.  Admins must explicitly opt-in to alpha features, and must be able to disable/prevent the use of beta features.

Reply:
@enj the feature you referenced seemed to go straight to Beta. Would that also be an option here? Meaning that we feature-gate the feature buy have it enabled by default?

Reply:
I have updated the KEP to add this as an alpha feature.